### PR TITLE
Show a pass/fail message for a completed identity verification

### PR DESCRIPTION
**Pull request by erikdstock**
This PR changes the IdentityVerificationApp (at `/identity-verification/:id`) to show different pages if the IDV is complete (that is has a state of 'passed' or 'failed'), blocking someone from beginning new flows at this point.

### CODE REVIEW

File: src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import { Box, Button, Link, Sans, Serif } from "@artsy/palette"
+import React from "react"
+
+export const CompleteFailed = () => {
```
Comment (damassi) [inline src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx]: `React.FC`

File: src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  return (
+    <AppContainer>
+      <HeadTitle>Artsy | ID Verification</HeadTitle>
+      <Meta
+        name="viewport"
+        content="width=device-width, initial-scale=1, maximum-scale=5 viewport-fit=cover"
+      />
+      <InnerComponent />}
```
Comment (damassi) [inline src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx]: Stray `}`

File: src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import { Box, Button, Link, Sans, Serif } from "@artsy/palette"
+import React from "react"
+
+export const CompleteFailed: React.FC = () => {
+  return (
+    <Box px={[2, 3]} mb={6} mt={4}>
+      <Box
+        mx={["auto"]}
```
Comment (yuki24) [inline src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx]:
```suggestion
        mx="auto"
```

File: src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        width={["100%", "80%"]}
+        maxWidth={"400px"}
```
Comment (yuki24) [inline src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx]:
```suggestion
        maxWidth="400px"
```

File: src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        textAlign="center"
+      >
+        <Serif size="6" color="black100">
```
Comment (yuki24) [inline src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx]:
```suggestion
        <Serif size="6">
```

File: src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import { Box, Button, Link, Sans, Serif } from "@artsy/palette"
+import React from "react"
+
+export const CompleteFailed: React.FC = () => {
+  return (
+    <Box px={[2, 3]} mb={6} mt={4}>
+      <Box
```
Comment (yuki24) [inline src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx]: I'm not sure if you have to nest `<Box>`es here. You could find [a refactoring example here](https://github.com/artsy/reaction/pull/3282/files#diff-b6e1258870cda8e1ddc88e37a1932561L129-R123).

File: src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+  const trackClickedContinueToVerification = () => {
+    trackEvent({
+      context_page_owner_id: identityVerification.internalID,
+      action_type: Schema.ActionType.ClickedContinueToIdVerification,
+      context_page: Schema.PageName.IdentityVerificationPage,
+    })
+  }
+
+  async function startIdentityVerification() {
```
Comment (yuki24) [inline src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx]: This function doesn't seem to use `await`. We should stick with using `async`/`await` together or not using them at all.

File: src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    )
+  }
+
+  return (
+    <AppContainer>
+      <HeadTitle>Artsy | ID Verification</HeadTitle>
+      <Meta
+        name="viewport"
+        content="width=device-width, initial-scale=1, maximum-scale=5 viewport-fit=cover"
+      />
```
Comment (yuki24) [inline src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx]: I don't think we need this `<Meta>` tag.

File: src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+import { RelayProp } from "react-relay"
+import { useTracking } from "react-tracking"
+import { commitMutation, graphql } from "relay-runtime"
+import createLogger from "Utils/logger"
+
+const logger = createLogger("StartIdentityVerification.tsx")
+
+interface Props {
+  relay: RelayProp
+  identityVerification: {
```
Comment (yuki24) [inline src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx]:
```suggestion
  identityVerification: IdentityVerificationApp_me['identityVerification']
```

File: src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    mutation
+      .then(handleMutationSuccess)
+      .catch(handleMutationError)
+      .finally(() => {
+        setRequesting(false)
+      })
+  }
+
+  const handleMutationSuccess = (identityVerificationFlowUrl: string) => {
+    location.assign(identityVerificationFlowUrl)
```
Comment (yuki24) [inline src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx]: We should probably use `props.router` or `props.location` rather than relying on the global one.

File: src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx (hunk 4 above)
Comment (dleve123) [reply src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx]: When are the `{}` needed? When this is evaluating an expression?

File: src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+            .
+          </Sans>
+        </Box>
+        <a href="mailto:verification@artsy.net">
+          <Button
+            block
+            width={["100%", 335]}
+            mt={4}
+            onClick={() => {
+              // trackClickedContinueToVerification()
```
Comment (dleve123) [inline src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx]: Should we delete this commented out code?

File: src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+      <ErrorModal
+        show={showErrorModal}
+        contactEmail="verification@artsy.net"
+        onClose={() => {
+          setShowErrorModal(false)
+        }}
+      />
+      <Box px={[2, 3]} mb={6} mt={4}>
+        <Box
+          mx={["auto"]}
```
Comment (dleve123) [inline src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx]: Similar to @yuki24's [comment above](https://github.com/artsy/reaction/pull/3294/files#r396483819), I wonder if `mx="auto"` is sufficient here.

File: src/Apps/IdentityVerification/__tests__/IdentityVerificationApp.test.tsx [hunk 13]
```diff
@@ -134,13 +173,13 @@ describe("IdentityVerification route", () => {
...
         env.mutations.mockNetworkFailureOnce()
 
-        await page.clickStartVerification()
-
-        const errorModal = expectOne(page.find(ErrorModal))
-        expect(errorModal.props().show).toBe(true)
-        expect(page.text()).toContain(
-          "Something went wrong. Please try again or contact verification@artsy.net."
-        )
+        page.clickStartVerification().then(() => {
```
Comment (dleve123) [inline src/Apps/IdentityVerification/__tests__/IdentityVerificationApp.test.tsx]: For my own education – why the change the `await` to `then` here? Does this have to do with the Jest error that you tracked down? Is it just for consistency purposes?

**Comment by yuki24** [comment]
I think this PR needs a bit more work to be ready as it requires a sizable change (global variable leak, typing, nested components, etc). I'd be also happy to pair on this if the comments here are overwhelming.

File: src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx (hunk 4 above)
Comment (erikdstock) [reply src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx]: Strings work by default, but for other js types including numbers you need `{}`. In some cases they get left behind when a responsive value for palette/styled-system (in the form of an array) gets removed - with the single string they are no longer necessary.

File: src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx (hunk 9 above)
Comment (erikdstock) [reply src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx]: stylistic choice - I'm fine with doing that but it felt more react-y to just pass the props i need since this component is not a fragment container and knows nothing about relay (except for having a prop named `relay` of course). I'm fine either way, happy to move it.

File: src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx (hunk 4 above)
Comment (damassi) [reply src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx]: We shouldn't be using `{}` unless interpolatred JS is actually being used.

File: src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx (hunk 3 above)
Comment (damassi) [reply src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx]: Saw that this was resolvrd, but @yuki24's requested changes wern't applied.

Comment (erikdstock) [reply src/Apps/IdentityVerification/IdentityVerificationApp/CompleteFailed.tsx]: yes, i have them all in a working branch, sorry - will change pr to WIP

File: src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+      <InnerComponent />
+    </AppContainer>
+  )
+}
+
+export const IdentityVerificationAppFragmentContainer = createFragmentContainer(
```
Comment (yuki24) [inline src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx]: Now that I look at this component again I think we should co-locate this fragment and `StartIdentityVerification` since this `IdentityVerificationApp` component doesn't directly depend on any GraphQL data.

File: src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx (hunk 9 above)
Comment (erikdstock) [reply src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx]: I updated the props to be only `relayEnvironment` and `identityVerificationId`.

File: src/Apps/IdentityVerification/__tests__/IdentityVerificationApp.test.tsx (hunk 13 above)
Comment (erikdstock) [reply src/Apps/IdentityVerification/__tests__/IdentityVerificationApp.test.tsx]: Yes, i think this was while i was trying to debug errors in our tests. As it turns out this didn't fix the issue so i'll fix now.

File: src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx (hunk 14 above)
Comment (erikdstock) [reply src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx]: It actually does depend on the state and that the `me` owns the `identityVerification` (their ids match). So i could make another fragment container but the amount of threading seems relatively small for the added complexity there.

File: src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx (hunk 9 above)
Comment (erikdstock) [reply src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx]: (and now `router`)

File: src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+            onClick={() => {
+              setRequesting(true)
+              trackClickedContinueToVerification()
+              startIdentityVerification()
+            }}
+          >
+            Continue to verification
+          </Button>
+        </Box>
+      </Box>
```
Comment (yuki24) [inline src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx]: I think we should bring back what is currently in the `<IdentityVerificationApp>` component. It seems like the old code was copy-&-pasted here, and it isn't just un-refactored code but also missing a link to FAQ.

https://github.com/yuki24/reaction/blob/bfffe3297b704f10378fa7d7831bac05427659da/src/Apps/IdentityVerification/IdentityVerificationApp.tsx#L117-L169

File: src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+interface Props {
+  me: IdentityVerificationApp_me
+  relay: RelayProp
+  router: Router
+}
+
+const IdentityVerificationApp: React.FC<Props> = ({ me, relay, router }) => {
+  const { identityVerification } = me
+
+  if (!identityVerification || identityVerification.userID !== me.internalID) {
```
Comment (yuki24) [inline src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx]: You don't have to change this as part of this PR, but It's odd that you have to do this on the front-end when the `identityVerification` field is a nested field of `me`. The front-end should only check the presence of `identityVerification` and the back-end should take care of hiding `identityVerification`s that belongs to someone else.

File: src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+export const IdentityVerificationAppFragmentContainer = createFragmentContainer(
+  IdentityVerificationApp,
+  {
+    me: graphql`
+      fragment IdentityVerificationApp_me on Me
+        @argumentDefinitions(id: { type: "String!" }) {
+        internalID
+        name
```
Comment (yuki24) [inline src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx]: We don't seems to be using it:

```suggestion
```

File: src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx (hunk 16 above)
Comment (erikdstock) [reply src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx]: I agree with you technically - but we had many long conversations and about this approach last month and I don't think it is going to change; a `me.identityVerification` is an idv that the user has access to, not necessarily that _belongs to_ the user, so we do need this check on the client.

File: src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx (hunk 15 above)
Comment (erikdstock) [reply src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx]: Good catch! probably lost on rebase. I'll try restoring this version but personally don't like the mixture of inner components.

Comment (erikdstock) [reply src/Apps/IdentityVerification/IdentityVerificationApp/StartIdentityVerification.tsx]: replaced :+1:

File: src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx [hunk 18]
```diff
@@ -113,60 +119,66 @@ const IdentityVerificationApp: React.FC<Props> = ({ me, relay }) => {
...
+              Artsy identity verification
+            </Serif>
+
+            <Sans size="4" mt="2" weight="medium">
+              You’ll need
+            </Sans>
+            <Sans size="4">
+              • A camera on your phone or computer
+              <br />
+              respont • Your government ID
```
Comment (yuki24) [inline src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx]: Maybe this might have slipped in by accident?
```suggestion
             • Your government ID
```

File: src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx (hunk 16 above)
Comment (dleve123) [reply src/Apps/IdentityVerification/IdentityVerificationApp/index.tsx]: If I am remembering context correctly, it's worth noting that this is _just_ to protect against the case of an admin accessing this page for an IDV that isn't theirs. At an API layer, we don't allow regular user A to query for the identity verifications for regular user B. In that case, the identity_verification would be `null` (the LHS of this expression).

cc @yuki24

**Comment by artsyit** [comment]
:rocket: PR was released in `v25.35.3` :rocket:
